Basic Module Support

freeformz opened the pull request:
The plan is to start here and add test cases/fixes for other uses as things progress.

Commit:
Start with some failing tests

Commit:
Need a way to specify heroku build info

This is inspired by build tags, but I'm not sure I like this.

The idea is that it could be used like this:

// +heroku <key1> <value1> [<value2>...]

I believe all the existing options translate...

// +heroku goVersion go1.11
// +heroku install ./cmd/... ./foo
// +heroku additionalTools github.com/mattes/migrate some.other/package

I tested a bit and `go mod tidy` preserves comments, so I think the
comments would be safe from automated tooling nuking them.

This format also makes extraction super easy:

$ awk '{ if ($1 == "//" && $2 == "+heroku" && $3 == "goVersion" ) { print $4 } }' ./go.mod
go1.11

or

$ awk '{ if ($1 == "//" && $2 == "+heroku" && $3 == "install" ) { print substr($0, index($0,$4)) }}' ./go.mod
./cmd/... ./foo

Commit:
Detect go.mod files

Commit:
Fix version / missing go version for tests

Commit:
Support basic build

This was incredibly easy.

Commit:
Support default package spec message for go modules

Commit:
Update changelog and README

owenthereal commented on lib/common.sh:
```diff
@@ -267,7 +268,11 @@ setGoVersionFromEnvironment() {
-    if [ -f "${depTOML}" ]; then
+    if [ -f "${goMOD}" ]; then
+        TOOL="gomodules"
+        ver=${GOVERSION:-$(awk '{ if ($1 == "//" && $2 == "+heroku" && $3 == "goVersion" ) { print $4 } }' ${goMOD})}
```
Should we exit `awk` after the first match? If users mistaken have multiple `// +heroku goVersion 1.10` there will be multiple returns:

```
awk '{ if ($1 == "//" && $2 == "+heroku" && $3 == "goVersion" ) { print $4; exit; } }' ${goMOD}
```

freeformz replied:
Good catch.

owenthereal commented on bin/compile:
```diff
@@ -400,6 +407,21 @@ export GOPATH
 # GB installation
 case "${TOOL}" in
+    gomodules)
+        pkgs=${GO_INSTALL_PACKAGE_SPEC:-$(awk '{ if ($1 == "//" && $2 == "+heroku" && $3 == "install" ) { print substr($0, index($0,$4)) }}' ${goMOD})}
```
Should we exit `awk` after the first match? If users mistaken have multiple lines of `// +heroku install ./cmd/...` there will be multiple returns:

```
awk '{ if ($1 == "//" && $2 == "+heroku" && $3 == "install" ) { print substr($0, index($0,$4)); exit; }}' ${goMOD}
```

owenthereal commented on test/run:
```diff
@@ -1,6 +1,46 @@
+  assertCaptured "github.com/heroku/fixture"
+  assertCapturedSuccess
+  assertCompiledBinaryExists
+}
```
Would it be possible to add a test for `// +heroku install xxx`?

freeformz replied:
Done.

owenthereal commented on lib/common.sh:
```diff
@@ -267,7 +268,11 @@ setGoVersionFromEnvironment() {
 
 determineTool() {
-    if [ -f "${depTOML}" ]; then
+    if [ -f "${goMOD}" ]; then
```
Do we also need to detect if `GO111MODULE=on` to enable gomodules?

freeformz replied:
I'd rather not until and issue is raised or we understand better the circumstances around needing to force it on.

owenthereal commented on test/fixtures/mod-deps/main.go:
```diff
@@ -0,0 +1,11 @@
+import (
+	"fmt"
+
+	"github.com/gorilla/mux"
```
Hmm...interesting. How did this test pass? Without specifying GO111MODULE=on, Go1.11rc2 doesn't fetch deps for me.

freeformz replied:
Are you using modules inside of a GOPATH? If so that's why. AFAICT, outside of a GOPATH, modules are turned on.

owenthereal replied:
Ah, that's true. I just verified it. I'm not sure I follow the rationale behind this though (I would expect it's either on or off).

So I assume GOPATH hasn't been set in our case that's why the test passes?

freeformz replied:
Yes, when a go.mod file is detected, we don't construct a $GOPATH, we just figure some stuff out and build the code. If this hold true I can't wait for the day when I can remove all the other tools. :-)

owenthereal replied:
❤️

Commit:
Missed bump to rc2 for testing purposes.

Commit:
Add a test for // +heroku install

Commit:
Ensure awk exits after the first match

Previously awk could output more than one line if a user specified more
than one install or goVersion heroku build directive.

Commit:
Don't fetch deps when -d vendor and len(go.sum) > 0

For those that wish to have repeatable builds in the face of Github
(or other sources) outages and/or private repositories that are vendored
enforce -mod=vendor if the `vendor` directory exists AND the size of
`go.sum` is > 0.

Either don't use `vendor` or don't commit `go.sum` if you want to skip
downloading and verification of dependencies.

I suspect that once modules are taken out of their experimental phase
we'll see a decrease in people using vendoring at all and only those who
need repeatable builds (or otherwise find value in them) will opt to
continue using it.
